src/PrimordialChem/test_primordial_chem.cpp

```diff
@@ -0,0 +1,171 @@
+{
+	// set initial conditions
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> dx = grid_elem.dx_;
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> prob_lo = grid_elem.prob_lo_;
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> prob_hi = grid_elem.prob_hi_;
+	const amrex::Box &indexRange = grid_elem.indexRange_;
+	const amrex::Array4<double> &state_cc = grid_elem.array_;
+
+	Real const Lx = (prob_hi[0] - prob_lo[0]);
+	Real const Ly = (prob_hi[1] - prob_lo[1]);
+	Real const Lz = (prob_hi[2] - prob_lo[2]);
+
+	// loop over the grid and set the initial condition
+	amrex::ParallelFor(indexRange, [=] AMREX_GPU_DEVICE(int i, int j, int k) {
+		Real const x = prob_lo[0] + (i + Real(0.5)) * dx[0];
+		Real const y = prob_lo[1] + (j + Real(0.5)) * dx[1];
+		Real const z = prob_lo[2] + (k + Real(0.5)) * dx[2];
+
+		Real temp = 0;
+		Real elec = 0;
+		Real hp = 0;
+		Real h = 0;
+		Real hm = 0;
+		Real dp = 0;
+		Real d = 0;
+		Real h2p = 0;
+		Real dm = 0;
+		Real h2 = 0;
+		Real hdp = 0;
+		Real hd = 0;
+		Real hepp = 0;
+		Real hep = 0;
+		Real he = 0;
+
+		amrex::ParmParse pp("primordialchem");
+		pp.query("temp", temp);
+		pp.query("elec", elec);
+		pp.query("hp", hp);
+		pp.query("h", h);
+		pp.query("hm", hm);
+		pp.query("dp", dp);
+		pp.query("d", d);
+		pp.query("h2p", h2p);
+		pp.query("dm", dm);
+		pp.query("h2", h2);
+		pp.query("hdp", hdp);
+		pp.query("hd", hd);
+		pp.query("hepp", hepp);
+		pp.query("hep", hep);
+		pp.query("he", he);
+
+		Real rho = 0; // g cm^-3
+
+		Real xmom = 0;
+		Real ymom = 0;
+		Real zmom = 0;
+		Real const P = 4.0e4 * quokka::boltzmann_constant_cgs; // erg cm^-3
+		Real Eint = (quokka::EOS_Traits<PrimordialChemTest>::gamma - 1.) * P;
+
+		Real const Egas = RadSystem<PrimordialChemTest>::ComputeEgasFromEint(rho, xmom, ymom, zmom, Eint);
+
+		state_cc(i, j, k, RadSystem<PrimordialChemTest>::gasEnergy_index) = Egas;
+		state_cc(i, j, k, RadSystem<PrimordialChemTest>::gasInternalEnergy_index) = Eint;
+		state_cc(i, j, k, RadSystem<PrimordialChemTest>::gasDensity_index) = rho;
+		state_cc(i, j, k, RadSystem<PrimordialChemTest>::x1GasMomentum_index) = xmom;
+		state_cc(i, j, k, RadSystem<PrimordialChemTest>::x2GasMomentum_index) = ymom;
+		state_cc(i, j, k, RadSystem<PrimordialChemTest>::x3GasMomentum_index) = zmom;
+
+		state_cc(i, j, k, HydroSystem<PrimordialChemTest>::scalar0_index) = elec;
+		state_cc(i, j, k, HydroSystem<PrimordialChemTest>::scalar0_index + 1) = hp;
+	});
+}
+
+auto problem_main() -> int
+{
+	// Problem parameters
+	const double CFL_number = 0.25;
+	const double max_time = 5e16; // > 1 Gyr
+	const int max_timesteps = 2e4;
+
+	// Problem initialization
+	constexpr int ncomp_cc = Physics_Indices<PrimordialChemTest>::nvarTotal_cc;
+	amrex::Vector<amrex::BCRec> BCs_cc(ncomp_cc);
+	for (int n = 0; n < ncomp_cc; ++n) {
+		BCs_cc[n].setLo(1, amrex::BCType::foextrap); // extrapolate
+		BCs_cc[n].setHi(1, amrex::BCType::foextrap);
+		BCs_cc[n].setLo(1, amrex::BCType::foextrap);
```

Comment:
This does not set the boundary conditions correctly. The first argument here is the dimension (0, 1, or 2).